### PR TITLE
Fix SDR classifier Region

### PR DESCRIPTION
**Pull request by ywcui1990**
Fixes: https://github.com/numenta/nupic/issues/3651
Summary
* handle multiple categories in sdr classifier
* call sdr classifier once in sdr classifier region
* add unit test of sdr classifier class to test multiple category prediction
* add integration test of sdr classifier region

@scottpurdy This is a fix for the broken SDR classifier. If this looks good to you, I will make the same change in nupic.core

### CODE REVIEW

File: src/nupic/algorithms/sdr_classifier.py [hunk 1]
```diff
@@ -204,8 +204,15 @@ def compute(self, recordNum, patternNZ, classification, learn, infer):
       print "  patternNZ (%d):" % len(patternNZ), patternNZ
       print "  classificationIn:", classification
 
-    # Store pattern in our history
-    self._patternNZHistory.append((recordNum, patternNZ))
+    # ensures that recordNum increases monotonically
+    if len(self._patternNZHistory) > 0:
+      if recordNum < self._patternNZHistory[-1][0]:
+        raise ValueError("the record number has to increase monotonically")
```
Comment (rhyolight) [inline src/nupic/algorithms/sdr_classifier.py]: Add to the docstring:

```
:raises: (ValueError) when record number does not increase monotonically.
```

**Comment by scottpurdy** [comment]
@ywcui1990 - are you ready to merge this? I want to revalidate #3665 once this is merged to make sure the two changes are compatible.

**Comment by ywcui1990** [comment]
@scottpurdy I am ready to merge this. I will push a separate PR in nupic.core.  I expect this PR to break regression tests since it affects classification/prediction accuracy. Do you know why the regression tests pass with this PR?

**Comment by scottpurdy** [comment]
I believe the regression tests only run on a nightly for master, not PRs. @rhyolight - is that right?

**Comment by rhyolight** [comment]
@scottpurdy Yes, it's nightly on master only.

**Pull request merged**